src/CodeRunner.ts

```diff
@@ -0,0 +1,277 @@
+     */
+    readonly inputManager: InputManager;
+    /**
+     * The backend that executes the code asynchronously
+     */
+    private backend: Promise<SyncClient<Backend>>;
+    /**
+     * Current state of the program
+     */
+    private state: RunState;
+    /**
+     * Buttons managed by this component
+     */
+    private buttons: Array<DynamicButton>;
+
+    /**
+     * Construct a new RunStateManager with the given listeners
+     * @param {ProgrammingLanguage} programmingLanguage The language to use
+     */
+    constructor(programmingLanguage: ProgrammingLanguage) {
+        this.programmingLanguage = programmingLanguage;
+        this.editor = new CodeEditor();
+        this.inputManager = new InputManager(async (input: string) => {
+            (await this.backend).writeMessage(input);
+            this.setState(RunState.Running);
+        });
+        this.backend = Promise.resolve({} as SyncClient<Backend>);
+        this.buttons = [];
+        this.addButton({
+            id: RUN_BTN_ID,
+            buttonText: t("Papyros.run"),
+            extraClasses: "text-white bg-blue-500"
+        }, () => this.runCode());
+        this.addButton({
+            id: STOP_BTN_ID,
+            buttonText: t("Papyros.stop"),
+            extraClasses: "text-white bg-red-500"
+        }, () => this.stop());
+        BackendManager.subscribe(BackendEventType.Input,
+            () => this.setState(RunState.AwaitingInput));
+        this.state = RunState.Ready;
+    }
+
+    /**
+     * Start the backend to enable running code
+     */
+    async start(): Promise<void> {
+        this.setState(RunState.Loading);
+        const backend = BackendManager.startBackend(this.programmingLanguage);
+        this.editor.setLanguage(this.programmingLanguage);
+        // Use a Promise to immediately enable running while downloading
+        this.backend = new Promise(resolve => {
+            return backend.workerProxy
+                // Allow passing messages between worker and main thread
+                .launch(proxy((e: BackendEvent) => BackendManager.publish(e)))
+                .then(() => {
+                    this.editor.setCompletionSource(async context => {
+                        const completionContext = Backend.convertCompletionContext(context);
+                        return backend.workerProxy.autocomplete(completionContext);
+                    });
+                    return resolve(backend);
+                });
+        });
+        this.editor.focus();
+        this.setState(RunState.Ready);
+    }
+
+    /**
+     * Interrupt the currently running code
+     * @return {Promise<void>} Promise of stopping
+     */
+    async stop(): Promise<void> {
+        this.setState(RunState.Stopping);
+        BackendManager.publish({
+            type: BackendEventType.End,
+            data: "User cancelled run", contentType: "text/plain"
+        });
+        await this.backend.then(b => b.interrupt());
```

Comment:
This is still assuming that nothing will catch the KeyboardInterrupt. You can only do this if `call` throws `InterruptError`. Treat client.interrupt as a suggestion/request.

Reply:
The goal is to fully interrupt. Ideally, some method should exist in synClient to force-interrupt (maybe a configurable timeout that terminates the worker if the interrupt was not successful?) Having to interpret client.interrupt as a suggestion is counter-intuitive to me.